kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,447 @@
+/// The arguments to the callback are:
+/// * `context`: an `&mut context` argument. this can be anything that engine needs to pass through to each call
+/// * `CDFScanFile`: a [`CDFScanFile`] struct that holds all the metadata required to perform Change Data
+///   Feed
+///
+/// ## Context
+/// A note on the `context`. This can be any value the engine wants. This function takes ownership
+/// of the passed arg, but then returns it, so the engine can repeatedly call `visit_cdf_scan_files`
+/// with the same context.
+///
+/// ## Example
+/// ```ignore
+/// let mut context = [my context];
+/// for res in scan_data { // scan data table_changes_scan.scan_data()
+///     let (data, vector, remove_dv) = res?;
+///     context = delta_kernel::table_changes::scan_file::visit_cdf_scan_files(
+///        data.as_ref(),
+///        selection_vector,
+///        context,
+///        my_callback,
+///     )?;
+/// }
+/// ```
+#[allow(unused)]
+pub(crate) fn visit_cdf_scan_files<T>(
+    data: &dyn EngineData,
+    selection_vector: &[bool],
+    context: T,
+    callback: CDFScanCallback<T>,
+) -> DeltaResult<T> {
+    let mut visitor = CDFScanFileVisitor {
+        callback,
+        selection_vector,
+        context,
+    };
+
+    visitor.visit_rows_of(data)?;
+    Ok(visitor.context)
+}
+
+// add some visitor magic for engines
+#[allow(unused)]
+struct CDFScanFileVisitor<'a, T> {
+    callback: CDFScanCallback<T>,
+    selection_vector: &'a [bool],
+    context: T,
+}
+
+impl<T> RowVisitor for CDFScanFileVisitor<'_, T> {
+    fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
+        require!(
+            getters.len() == 18,
+            Error::InternalError(format!(
+                "Wrong number of CDFScanFileVisitor getters: {}",
+                getters.len()
+            ))
+        );
+        for row_index in 0..row_count {
+            if !self.selection_vector[row_index] {
+                continue;
+            }
+            let timestamp = getters[16].get(row_index, "scanFile.timestamp")?;
+            let commit_version: i64 = getters[17].get(row_index, "scanFile.commit_version")?;
+            let commit_version: u64 = commit_version.try_into().map_err(Error::generic)?;
+
+            if let Some(path) = getters[0].get_opt(row_index, "scanFile.add.path")? {
+                let deletion_vector = visit_deletion_vector_at(row_index, &getters[1..=5])?;
+                let dv_info = DvInfo { deletion_vector };
+                let partition_values =
+                    getters[6].get(row_index, "scanFile.add.fileConstantValues.partitionValues")?;
+                let scan_file = CDFScanFile {
+                    scan_type: CDFScanFileType::Add,
+                    path,
+                    dv_info,
+                    partition_values,
+                    commit_version,
+                    timestamp,
+                };
+                (self.callback)(&mut self.context, scan_file)
```

Comment:
nit: There is an annoying amount of redundancy between the three cases... annoying because there are just enough code differences that it may be the correct approach, rather than trying to factor out something complex. 

The alternative would be something like this:
```rust
            let (scan_type, path, deletion_vector, partition_values) =
                if let Some(path) = getters[0].get_opt(row_index, "scanFile.add.path")? {
                    let deletion_vector = visit_deletion_vector_at(row_index, &getters[1..=5])?;
                    let partition_values =
                        getters[6].get(row_index, "scanFile.add.fileConstantValues.partitionValues")?;
                    (CDFScanFileType::Add, path, deletion_vector, partition_values)
                } else if let Some(path) = getters[7].get_opt(row_index, "scanFile.remove.path")? {
                      ...
                } else if let Some(path) = getters[14].get_opt(row_index, "scanFile.cdc.path")? {
                      ...
                    (CDFScanFileType::Cdc, path, None, partition_values)
                } else {
                    continue;
                }
            
            let dv_info = DvInfo { deletion_vector };
            let scan_file = CDFScanFile {
                scan_type,
                path,
                dv_info,
                partition_values,
                commit_version: getters[17].get(row_index, "scanFile.commit_version")?,
                timestamp: getters[16].get(row_index, "scanFile.timestamp")?,
            };
            (self.callback)(&mut self.context, scan_file)
```
(saves ~15 LoC)

Reply:
Yeah I struggled with figuring a way to make it neater, but I really like this!